samples/SocialSample/Startup.cs

```diff
@@ -228,6 +231,99 @@ public void Configure(IApplicationBuilder app)
+            {
+                signinApp.Run(async context =>
+                {
+                    // Setting DefaultAuthenticateScheme causes User to be set
+                    // var user = context.User;
+
+                    // This is what [Authorize] calls
+                    var userResult = await context.AuthenticateAsync();
+                    var user = userResult.Principal;
+                    var authProperties = userResult.Properties;
+
+                    // This is what [Authorize(ActiveAuthenticationSchemes = MicrosoftAccountDefaults.AuthenticationScheme)] calls
+                    // var user = await context.AuthenticateAsync(MicrosoftAccountDefaults.AuthenticationScheme);
+
+                    // Deny anonymous request beyond this point.
+                    if (!userResult.Succeeded || user == null || !user.Identities.Any(identity => identity.IsAuthenticated))
+                    {
+                        // This is what [Authorize] calls
+                        // The cookie middleware will handle this and redirect to /login
+                        await context.ChallengeAsync();
+
+                        // This is what [Authorize(ActiveAuthenticationSchemes = MicrosoftAccountDefaults.AuthenticationScheme)] calls
+                        // await context.ChallengeAsync(MicrosoftAccountDefaults.AuthenticationScheme);
+
+                        return;
+                    }
+
+                    var refreshToken = authProperties.GetTokenValue("refresh_token");
+
+                    if (string.IsNullOrEmpty(refreshToken))
+                    {
+                        context.Response.ContentType = "text/html";
+                        await context.Response.WriteAsync("<html><body>");
+                        await context.Response.WriteAsync("No refresh_token is available.<br>");
+                        await context.Response.WriteAsync("<a href=\"/\">Home</a>");
+                        await context.Response.WriteAsync("</body></html>");
+                    }
+
+                    if (string.Equals(GoogleDefaults.AuthenticationScheme, user.Identities.First().AuthenticationType))
+                    {
+                        var handler = (AuthenticationHandler<GoogleOptions>)await authSchemes.GetHandlerAsync(context, GoogleDefaults.AuthenticationScheme);
+
+                        var pairs = new Dictionary<string, string>()
+                        {
+                            { "client_id", handler.Options.ClientId },
+                            { "client_secret", handler.Options.ClientSecret },
+                            { "grant_type", "refresh_token" },
+                            { "refresh_token", refreshToken }
+                        };
+                        var content = new FormUrlEncodedContent(pairs);
+                        var response = await handler.Options.Backchannel.PostAsync(GoogleDefaults.TokenEndpoint, content, context.RequestAborted);
```

Comment:
> It's not clear what the input or output of Renew would be.

How about something like:

```csharp
AuthenticateResult RenewAsync(AuthenticationProperties properties);
```

> if it would operate on the SignInScheme

IMO, it should be consistent with what we already do in the callback part, so yeah, it should probably call `SignInAsync(SignInScheme)` to return a new cookie with the updated identity/claims/tokens.